### PR TITLE
docs: document design decisions regarding Options structs in APIs

### PR DESCRIPTION
**Pull request by vangent**

### CODE REVIEW

File: internal/docs/design.md [hunk 1]
```diff
@@ -107,6 +107,36 @@ symbols stable over time.
...
+    constructor. Inside a driver interface or in a concrete type like `blob`,
+    use more descriptive names like `ReaderOptions` or `WriterOptions`.
+-   When similar `Options` are part of a driver interface and also part of the
+    concrete type (e.g., `blob.WriterOptions`), duplicate the struct instead of
+    aliasing or embedding it, and copy the struct fields explicitly where
+    needed. This allows the godoc for each type to be tailored to the
+    appropriate audience (concrete type: users, driver interface: provider
+    implementors), and also allows the structs to diverge over time if
+    appropriate.
+-   Arguments that are optional, including ones for which a default value is
```
Comment (zombiezen) [inline internal/docs/design.md]: I recommend wording this more strongly. Anything that is required should not be in an options struct. Adding a new field to the options struct must have a reasonable default, else it is a breaking change.

File: internal/docs/design.md [hunk 2]
```diff
@@ -107,6 +107,36 @@ symbols stable over time.
...
+    common, should generally go into the `Options` struct. Avoid adding
+    `Options` fields for which the default value doesn't make sense.
+-   Regarding empty `Options` structs: we considered only adding them the first
```
Comment (zombiezen) [inline internal/docs/design.md]: This prose feels like it should be part of the first paragraph when it is introducing the idea. The rest of the bullets work nicely to explain the guidelines efficiently.

File: internal/docs/design.md [hunk 3]
```diff
@@ -107,6 +107,36 @@ symbols stable over time.
...
+    possible that we'll add arguments, add `Options`.
+-   Name the `Options` struct appropriately. `Options` is usually fine for
+    provider constructors since the package generally only exposes a
+    constructor. Inside a driver interface or in a concrete type like `blob`,
+    use more descriptive names like `ReaderOptions` or `WriterOptions`.
+-   When similar `Options` are part of a driver interface and also part of the
+    concrete type (e.g., `blob.WriterOptions`), duplicate the struct instead of
+    aliasing or embedding it, and copy the struct fields explicitly where
+    needed. This allows the godoc for each type to be tailored to the
+    appropriate audience (concrete type: users, driver interface: provider
```
Comment (zombiezen) [inline internal/docs/design.md]: It took me a while to parse this aside. I suggest:

```suggestion
    appropriate audience (e.g. end-users for the concrete type, provider implementors for the driver interface)
```

File: internal/docs/design.md (hunk 2 above)
Comment (ijt) [inline internal/docs/design.md]: we considered only adding them the first -> we considered only adding them _when_ the first

File: internal/docs/design.md (hunk 1 above)
Comment (vangent) [reply internal/docs/design.md]: Done.

File: internal/docs/design.md (hunk 2 above)
Comment (vangent) [reply internal/docs/design.md]: It's a pretty long aside, and I think distracts from the main points. I pulled it out of the bullet list and left it at the end.

Comment (vangent) [reply internal/docs/design.md]: Done.

**Comment by vangent** [comment]
PTAL.

File: internal/docs/design.md [hunk 4]
```diff
@@ -107,6 +107,36 @@ symbols stable over time.
...
+-   When similar `Options` are part of a driver interface and also part of the
+    concrete type (e.g., `blob.WriterOptions`), duplicate the struct instead of
+    aliasing or embedding it, and copy the struct fields explicitly where
+    needed. This allows the godoc for each type to be tailored to the
+    appropriate audience (e.g. end-users for the concrete type, provider implementors for the driver interface)
+    implementors), and also allows the structs to diverge over time if
+    appropriate.
+-   Required arguments must not be in an `Options` struct.
+-   All fields of the `Options` struct must have reasonable defaults.
+
```
Comment (jba) [inline internal/docs/design.md]: Add:
- The argument should be of type `*Options`, so `nil` can be passed in the default case.

Comment (vangent) [reply internal/docs/design.md]: This is already submitted, but done in #884.

**Pull request merged**